lvm: cleanups

flokli opened the pull request:
This includes quite some fixes to lvm2 accumulated while working on enabling cryptsetup support in systemd (https://github.com/NixOS/nixpkgs/pull/66856). I intend to rebase #66856 on top of the changes here.

lvm2 is now a multi-output derivation, which helps for closure size reduction, but also becomes necessary as soon as you try to build systemd with cryptsetup support (that requires the libdevmapper component of lvm2) - due to lvm itself also containing references to libsystemd (via their shipped systemd generators).

Properly splitting this avoids getting rid of hacks like `systemd_with_lvm2`.

This also contains quite some cleanups in the unit itself, making it more maintainable in the longer run.

```
nixos/tests/installer: lvm: test lvm2-pvscan@ units
nixos/test/installer: add postBootCommands
systemd_with_lvm2: remove
nixos/tasks/lvm: add dmeventd and lvmthin support
lvm2: only split bin and lib out from out if cmdlib isn't enabled
lvm2: fix location to systemd-run invocation
lvm2: fix paths to use /run instead of /var/run.
lvm2: add multiple output support
lvm2: don't embed ./configure line in lvm2 binary
lvm2: make --enable-cmdlib optional
lvm2: 2.03.01 -> 2.03.09
lvm2: fetch sources from http instead of git
lvm2: add myself as maintainer
lvm2: cleanups, make udev optional
lvm2: remove unused default.upstream file
```

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

flokli committed:
systemd_with_lvm2: remove

This is already handled by the lvm module, which now properly adds the
lvm systemd generators.

Initially introduced in c0fd88748a66c9ca38092fdbf7e97f765c8e411d.

flokli committed:
lvm2: cleanups, make udev optional

 - use installTargets again ("install", and
   "install_systemd_{generators,units,configuration}" when udev is not
   null)
 - The call to the blkid binary in lvm2's 13-dm-disk.rules file
   disappeared (so we don't need to patch in blkid anymore).
   lvm seems to rely on udev's internal blkid functionality.
 - Call /run/current-system/systemd/bin/udevadm instead
   of ${systemd}/bin/udevadm in the lvm activation generator.
   This is not necessary to break the dependency cycle (as we don't use
   that file when building without udev), but a good idea anyways -
   We want to trigger the udevadm of the current system, not the one
   that lvm was built with.

flokli committed:
lvm2: fetch sources from http instead of git

Otherwise, we end up in a dependency cycle:

systemd -> cryptsetup -> lvm -> fetchgit -> git -> openssh -> libfido2 -> hidapi -> libusb -> udev=systemd

flokli committed:
lvm2: remove unused default.upstream file

ajs124 committed:
lvm2: fix paths to use /run instead of /var/run.

This shuts up some warnings.

ajs124 committed:
lvm2: add myself as maintainer

flokli committed:
lvm2: make --enable-cmdlib optional

This seems to be mostly used to simplify LV management tasks from a web
interface
(https://www.redhat.com/archives/linux-lvm/2008-September/msg00029.html),
and is as fat as the `lvm` binary itself

flokli committed:
lvm2: don't embed ./configure line in lvm2 binary

flokli committed:
lvm2: add multiple output support

ajs124 committed:
lvm2: 2.03.01 -> 2.03.09

ajs124 committed:
nixos/tasks/lvm: add dmeventd and lvmthin support

Introduce a pkgs.lvm2_dmeventd that contains dmeventd support, and
enable if services.lvm.dmeventd.enable is true.

ajs124 committed:
nixos/test/installer: add postBootCommands

ajs124 committed:
nixos/tests/installer: lvm: test lvm2-pvscan@ units

Also, add some sleep statements in between, which seems to at least feel
like it causes

> WARNING: Device /dev/vda* not initialized in udev database even after waiting 10000000 microseconds.

To occur less frequently.

This eventually still succeeds after some amount of waiting, I suspect
some racyness in the way lvm's udev-triggered scripts trigger other
units.

flokli committed:
lvm2: only split bin and lib out from out if cmdlib isn't enabled

flokli committed:
lvm2: fix location to systemd-run invocation

ajs124 commented on nixos/modules/tasks/lvm.nix:
```diff
@@ -1,17 +1,61 @@
   };
 
+  config = let
+    pkg = if cfg.dmeventd.enable then pkgs.lvm2_dmeventd else pkgs.lvm2;
```
Maybe this should be exposed as an (internal) option, in case someone has their own special lvm2 package with more units or something. Sure they could also just overlay the module, but that would mean a whole bunch of rebuilds.

flokli replied:
Good idea, exposed as an internal option.

ajs124 commented on pkgs/os-specific/linux/lvm2/default.nix:
```diff
@@ -1,48 +1,71 @@
-    rev = "v${builtins.replaceStrings [ "." ] [ "_" ] version}";
-    sha256 = "0jlaswf1srdxiqpgpp97j950ddjds8z0kr4pbwmal2za2blrgvbl";
+  src = fetchurl {
+    url = "https://mirrors.kernel.org/sourceware/lvm2/LVM2.${version}.tgz";
```
Couldn't this use `mirror://kernel` or something like that?

flokli replied:
This goes to cdn.kernel.org/pub, and I couldn't find it from there, which is why I left it like that.

ajs124 reviewed (approved):
Two minor suggestions, but otherwise this LGTM.

bjornfor commented on pkgs/os-specific/linux/lvm2/default.nix:
```diff
@@ -64,30 +87,39 @@ stdenv.mkDerivation {
   doCheck = false; # requires root
 
+  makeFlags = stdenv.lib.optionals (udev != null) [
+    "SYSTEMD_GENERATOR_DIR=$(out)/lib/systemd/system-generators"
```
Use placeholder instead of shell variable for $out?

flokli replied:
I think using shell variables is preferred over placeholder:
https://discourse.nixos.org/t/what-is-the-difference-between-placeholder-out-and-out/4862/2

flokli replied:
We use placeholder to not have to manually assemble `cmakeFlags`, but in places where using shell variables work, let's make use of it.

Mic92 commented on nixos/modules/tasks/lvm.nix:
```diff
@@ -1,17 +1,70 @@
+          mkdir -p /etc/lvm
+          cat << EOF >> /etc/lvm/lvm.conf
+          ${optionalString cfg.thin.enable ''
+            global/thin_check_executable = "$(which thin_check)"
```
```suggestion
            global/thin_check_executable = "$(command -v thin_check)"
```

flokli replied:
Updated this to use  `command -v`.

Mic92 commented on nixos/modules/tasks/lvm.nix:
```diff
@@ -1,17 +1,70 @@
+            global/thin_check_executable = "$(which thin_check)"
+          ''}
+          ${optionalString cfg.dmeventd.enable ''
+            dmeventd/executable = "$(which false)"
```
```suggestion
            dmeventd/executable = "$(command -v false)"
```

Mic92 commented on nixos/tests/installer.nix:
```diff
@@ -552,16 +555,26 @@ in {
           + " mkpart primary 2048M -1s"  # PV2
           + " set 2 lvm on",
           "udevadm settle",
+          "sleep 1",
```
Maybe add a comment why the sleep is necessary here.

flokli replied:
It's already in the commit message.

Also, it's not really "needed".

There's some racyness causing some waiting every now and then, and I at least felt like this will make it less likely to trigger. The test still succeeds, it's just less likely to take longer.

Mic92 commented on pkgs/os-specific/linux/lvm2/default.nix:
```diff
@@ -53,6 +53,9 @@ stdenv.mkDerivation rec {
       --replace /usr/bin/udevadm /run/current-system/systemd/bin/udevadm
+    # https://github.com/lvmteam/lvm2/issues/36
+    substituteInPlace udev/69-dm-lvm-metad.rules.in \
+      --replace "(BINDIR)/systemd-run" /run/current-system/systemd/bin/systemd-run
```
Is there no `$` before `(BINDIR)`?

arianvp replied:
no this is correct

arianvp replied:
https://github.com/lvmteam/lvm2/blob/master/udev/69-dm-lvm-metad.rules.in#L111

flokli replied:
No, that's some custom substitution mechanism - see https://github.com/lvmteam/lvm2/blob/master/udev/69-dm-lvm-metad.rules.in#L111 linked in the issue referenced above.

Mic92 commented on pkgs/os-specific/linux/lvm2/default.nix:
```diff
@@ -1,48 +1,71 @@
+, libuuid
 , thin-provisioning-tools, libaio
-, enable_dmeventd ? false }:
+, enable_cmdlib ? false
+, enable_dmeventd ? false
```
camelCase seems to be the way to go for enable options:

```suggestion
, enableCmdlib ? false
, enableDmeventd ? false
```

flokli replied:
okay, will update.

flokli replied:
At least `enable_dmeventd` has already existed before that, so I'll add a commit on top of that and not bother with the rebase.

flokli committed:
lvm2: rename enable_{cmdlib,dmeventd} to enable{Cmdlib,Dmeventd}

Make this more consistent with how these flags look like in the rest of
nixpkgs.

flokli commented:
Adressed @Mic92's comments.

arianvp commented:
LGTM

flokli committed:
lvm2: passthru tests

flokli commented:
Well, tests weren't in passthru. I pushed a commit adding them, good idea!

flokli commented:
Tests succeeded on ofborg x86_64-linux, and failed on aarch64 due to another unrelated error/flakyness:

```
 --- Error --- nix-build
error: --- SysError --- nix-daemon
```

andersk commented:
This broke multipath-tools; the fix in #93949 is needed.
